src/Aspire.Hosting/ResourceBuilderExtensions.cs

```diff
@@ -607,6 +608,14 @@ public static IResourceBuilder<T> WaitFor<T>(this IResourceBuilder<T> builder, I
                     $"Resource '{dependency.Resource.Name}' has entered the '{snapshot.State.Text}' state prematurely."
                     );
             }
+
+            // If our dependency resource has health check annotations we want to wait until they turn healthy
+            // otherwise we don't care about their health status.
+            if (dependency.Resource.TryGetAnnotationsOfType<HealthCheckAnnotation>(out var _))
```

Comment:
This should be in the initial call, not a call after.  This does not handle terminal states.

Reply:
I now remember why I did it this way :)

One of the quirks of the resource notification service is that the `Snapshot.Status` conflates resource visibility in the dashboard with status. So when the publisher calls `PublishAsync` for the resource state its going to a resource which is in a hidden state - so simultaneously looking for a running resource and a healthy resource is only possible if in the publisher we look up the replica IDs and publish directly to them.

Reply:
This is because of how WaitForResource is implemented right? Maybe skip hidden resources as part of those checks?

Reply:
Got child resource waits working.